src/Engine/ProtoCore/RuntimeData.cs

```diff
@@ -164,22 +178,24 @@ public CallSite GetCallSite(int classScope, string methodName, Executable execut
                     return nodeMap;
             }
 
+            List<string> callsiteIDs;
             foreach (Guid nodeGuid in nodeGuids)
             {
-                // Get a list of GraphNode objects that correspond to this node.
-                var matchingGraphNodes = graphNodes.Where(gn => gn.guid == nodeGuid);
-
-                if (!matchingGraphNodes.Any())
-                    continue;
-
-                // Get all callsites that match the graph node ids.
-                var matchingCallSites = (from cs in CallsiteCache
-                                         from gn in matchingGraphNodes
-                                         where string.Equals(cs.Key, gn.CallsiteIdentifier)
-                                         select cs.Value);
-
-                // Append each callsite element under node element.
-                nodeMap[nodeGuid] = matchingCallSites.ToList();
+                if (NodeToCallsiteIdentifiersMap.TryGetValue(nodeGuid, out callsiteIDs))
+                {
+                    var matchingCallSites = new List<CallSite>();
+                    foreach (var id in callsiteIDs)
+                    {
+                        CallSite callsite;
+                        if (CallsiteCache.TryGetValue(id, out callsite))
+                            matchingCallSites.Add(callsite);
+                    }
+                    nodeMap[nodeGuid] = matchingCallSites;
+                }
+                else
+                {
+                    nodeMap[nodeGuid] = new List<CallSite>();
+                }
```

Comment:
Then if we have the `NodeToCallsiteMap`, we won't need to do any of this either; we could simply return that map; or am I missing something?

Reply:
Yes... this function still might need to filter the map based on the incoming dynamo graph node Guids but it could be simpler.  Back in the D4R they could simply use the map